File: algorithms/DP/L1137_N-th_Tribonacci_Number.go
// Problem Source: https://leetcode.com/problems/n-th-tribonacci-number/
package main

import (
    "fmt"
    "os"
)

func tribonacci(n int) int {
    cache := make([]int, 3, n + 3)
    cache[0], cache[1], cache[2] = 0, 1, 1
    for i := 3; i <= n; i++ {
        cache = append(cache, cache[i - 1] + cache[i - 2] + cache[i - 3])
    }
    return cache[n]
}

type Testing struct {
    n, expect   int
}

func main() {
    testCases := []*Testing{
        { n: 4, expect: 4 },
        { n: 25, expect: 1389537 },
    }

    for _, t := range testCases {
        if outp := tribonacci(t.n); outp != t.expect {
            fmt.Println("FAIL")
            fmt.Println("Want: ", t.expect)
            fmt.Println("Got: ", outp)
            os.Exit(1)
        }
    }

    fmt.Println("ok")
}

File: problem-archive/DP/L1137_Nth_Tribonacci_Number.js
/**
 * @param {number} n
 * @return {number}
 */
var tribonacci = function(n) {
    if (n < 1) {
        return 0;
    }

    let r0 = 0;
    let r1 = 1;
    let r2 = 1;
    let t0 = 0;
    let t1 = 0;
    for (let i = 2; i < n; i++) {
        t0 = r1;
        t1 = r2;
        r2 += r1 + r0;
        r0 = t0;
        r1 = t1;
    }
    return r2;
};

console.log(tribonacci(4));
console.log(tribonacci(25));
